WIP, DOC: add doc version selector

tylerjereddy opened the pull request:
I was asked to add a doc version selector for our pages; this PR attempts to do that.

~~It is loosely based on the skimage approach to doing this, but instead of more tightly
integrating with the doc template machinery I've effectively anchored it below
the search box at the moment.~~

There are probably a few issues I haven't though of, but seems ok as a draft approach
locally. Not sure about portability to older doc builds if there's a desire to rebuild for them
eventually.

TODO: 

- [ ] best way to test the javascript time-travelling without experimentally re-deploying docs online? esp. given the vastly different local "URL" vs. online; could unit test the javascript functions but ugh
- [x] move the version selector menu to the top left of the page, similar to CPython docs, because it is otherwise at the mercy of margin content length in left column below search box (where it is currently anchored)
- [ ] add support for travelling to online devdocs from local devdocs?
- [ ] add support for the different base url used by devdocs
- [ ] decide if we're going to aim to rebuild older doc versions online to allow bidirectional time-travels--how much work to "backport" the menu patch to historical docs?
- [ ] remove debug console code & clean up javascript indent / formatting when ready to merge
- [ ] remove extraneous content remaining from the CPython docs javascript used as the foundation, and probably credit them for foundation code
- [ ] the drop-down version selector menu is currently way too wide

tylerjereddy committed:
DOC: add doc version selector.

tylerjereddy commented on doc/source/_templates/searchbox.html:
```diff
@@ -20,4 +20,30 @@ <h4>{{ _('Quick search') }}</h4>
 </div>
 <script type="text/javascript">$('#searchbox').show(0);</script>
+<div class="well">
+<strong>Docs for {{ release }}<br></strong>
```
I think this line produced a strange commit hash locally.

tylerjereddy commented on doc/source/_templates/searchbox.html:
```diff
@@ -20,4 +20,30 @@ <h4>{{ _('Quick search') }}</h4>
+                    base_url = 'https://www.numpy.org/devdocs/index.html'
+                }
+                else {
+                    base_url = 'https://docs.scipy.org/doc/numpy-VERSION'
```
More ambitious might be to try to preserve a users' location within the docs when "time travelling" to a different doc version (`np.sqrt` in 1.15 -> `np.sqrt` in 1.13); that said, skimage doesn't do this -- you just end up back at the index of the version you selected regardless of where you were before.

Might be neat, but also would be sensitive to doc restructuring over time causing breakages.

mattip replied:
Python itself has a [javascript](https://github.com/python/cpython/blob/master/Doc/tools/static/switchers.js) that does this pretty well. It builds a list of possible urls, and uses the first valid one. You can see it in action if you go to [docs.python.org](https://docs.python.org/3/reference/datamodel.html#objects-values-and-types) and click the version button at the top of the page

eric-wieser commented on doc/source/_templates/searchbox.html:
```diff
@@ -20,4 +20,30 @@ <h4>{{ _('Quick search') }}</h4>
+            var versions = ['dev', '1.15.1', '1.14.5', '1.13.0', '1.12.0', '1.11.0', '1.10.1', '1.9.3', '1.8.1', '1.7.0'];
+            for (i = 0; i < versions.length; i++) {
+
+                if (i == 0) {
```
Better to switch on the value of dev, not of i
```
versions.forEach(function(version) {
    if (version == 'dev') {
    }
    else {
    }
});
```

eric-wieser commented on doc/source/_templates/searchbox.html:
```diff
@@ -20,4 +20,30 @@ <h4>{{ _('Quick search') }}</h4>
+                    .replace('URL', base_url)
+                    .replace(/VERSION/g, versions[i]));
+            }
+        </script>
```
Indentation is very weird throughout this javacsript

tylerjereddy committed:
Initial / draft effort to switch over to CPython-style version selector

tylerjereddy committed:
DOC: commit in state where time travelling from one page to matching page in history is possible. [ci skip]

tylerjereddy commented:
Revised substantially to more closely match the standard Python doc version selector to allow for "time travelling while keeping your spot bookmarked." For example, quickly navigate from a local devdocs build `np.sqrt` to an older version of `np.sqrt` docs online using drop-down menu (scikit-image docs can't do that).

This is very rough (though it does work well-enough in my hands from base functionality standpoint) for now & needs a lot of work and careful testing / debugging of the javascript, so I'll leave the console logging in & add a to-do list below:

- [ ] best way to test the javascript time-travelling without experimentally re-deploying docs online? esp. given the vastly different local "URL" vs. online; could unit test the javascript functions but ugh
- [ ] move the version selector menu to the top left of the page, similar to CPython docs, because it is otherwise at the mercy of margin content length in left column below search box (where it is currently anchored)
- [ ] add support for travelling to online devdocs from local devdocs?
- [ ] add support for the different base url used by devdocs
- [ ] decide if we're going to aim to rebuild older doc versions online to allow bidirectional time-travels--how much work to "backport" the menu patch to historical docs?
- [ ] remove debug console code & clean up javascript indent / formatting when ready to merge
- [ ] remove extraneous content remaining from the CPython docs javascript used as the foundation

tylerjereddy commented on doc/source/_templates/searchbox.html:
```diff
@@ -21,3 +21,200 @@ <h4>{{ _('Quick search') }}</h4>
+    '(?:\\d)',
+    '(?:\\d\\.\\d[\\w\\d\\.]*)',
+    '(?:dev)',
+    '(?:release/\\d.\\d[\\x\\d\\.]*)'];
```
I moved away from using regexs (the CPython docs strategy) because I didn't want 2 problems instead of 1, at least for now.

tylerjereddy committed:
DOC: move menu to top left [ci skip]

* migrated the NumPy version selector
menu from a location anchored below the
search box to the "top left" of the web
page layout, similar to the layout used
in standard Python docs

* main motivation was to reduce the impact
of long left-side sidebars bumping the menu
down depending on location within docs

* best way I could find to do this was by
placing the menu javascript code inside a
rootrellink block

tylerjereddy committed:
DOC: some cleanup for version selector [ci skip]

* removed some of the left-over template code copied
in from CPython docs; especially material related
to unused regex code & language switcher, which doesn't
apply to NumPy

tylerjereddy committed:
DOC: more language and regex cleanup

tylerjereddy commented:
I spent a few hours trying to shorten the width of the version selector in the top left corner :/ Probably some sphinx / CSS interaction thing I'm missing.

I'll keep at it later this weekend maybe.

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,6 +1,134 @@
+    '1.9.3': '1.9.3',
+    '1.8.1': '1.8.1',
+    '1.7.0': '1.7.0',
+  };
```
Why a dictionary with identical keys and values?

tylerjereddy replied:
Inherited from [CPython Source](https://github.com/python/cpython/blob/master/Doc/tools/static/switchers.js#L12); might be able to simplify if it stays that way, but maybe not if we decide to have a special handling for `dev`

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,6 +1,134 @@
+
+    buf.push('</select>');
+    return buf.join('');
+  }
```
Given you already have jquery here, this function is simpler as:

```
$('<select>').append(
    versions.map(function(v) {
        var opt = $('<option>').attr('value', v).text(title);
        if (v == current_version) opt.prop('selected' , true);
        return opt;
    })
)
```

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,6 +1,134 @@
+        console.log('local dev doc build detected');
+	return 'dev';
+    }
+    else if (url.includes('devdocs')) {
```
Note that this `includes` is still somewhat new, and 12% of users have browsers that don't support it: https://caniuse.com/#feat=es6-string-includes

tylerjereddy commented:
the width of the drop-down menu is still a problem -- if I paste the `select` block in an online html renderer it looks just fine, so maybe this is related to the sphinx template / header links stuff.

tylerjereddy commented:
I'm going to close this PR of mine -- a few people said they'd try to fix the formatting / html issues & never got around to it, and html / sphinx resizing / positioning issues are not something I'm passionate to deal with.

Shekharrajak commented:
Hi @tylerjereddy , I want to continue work on this and may be I should cherry pick your commits. Please let me know if you have any suggestions in mind.

Thanks!
